src/test/optimizer/dp_ccp_test.cpp

```diff
@@ -0,0 +1,103 @@
+#include "gtest/gtest.h"
+
+#include "cost_model/cost_model_logical.hpp"
+#include "expression/expression_functional.hpp"
+#include "logical_query_plan/mock_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "optimizer/dp_ccp.hpp"
+#include "optimizer/join_graph.hpp"
+#include "statistics/column_statistics.hpp"
+#include "statistics/table_statistics.hpp"
+#include "storage/storage_manager.hpp"
+#include "testing_assert.hpp"
+#include "utils/load_table.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+class DpCcpTest : public ::testing::Test {
+ public:
+  void SetUp() override {
+    cost_estimator = std::make_shared<CostModelLogical>();
+
+    const auto column_statistics_a_a = std::make_shared<ColumnStatistics<int32_t>>(0.0f, 10.0f, 1, 50);
+    const auto column_statistics_b_a = std::make_shared<ColumnStatistics<int32_t>>(0.0f, 10.0f, 40, 100);
+    const auto column_statistics_c_a = std::make_shared<ColumnStatistics<int32_t>>(0.0f, 10.0f, 1, 100);
+
+    const auto table_statistics_a = std::make_shared<TableStatistics>(
+        TableType::Data, 20, std::vector<std::shared_ptr<const BaseColumnStatistics>>{column_statistics_a_a});
+    const auto table_statistics_b = std::make_shared<TableStatistics>(
+        TableType::Data, 3, std::vector<std::shared_ptr<const BaseColumnStatistics>>{column_statistics_b_a});
+    const auto table_statistics_c = std::make_shared<TableStatistics>(
+        TableType::Data, 5, std::vector<std::shared_ptr<const BaseColumnStatistics>>{column_statistics_c_a});
+
+    node_a = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}}, "a");
+    node_a->set_statistics(table_statistics_a);
+    node_b = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}}, "b");
+    node_b->set_statistics(table_statistics_b);
+    node_c = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}}, "c");
+    node_c->set_statistics(table_statistics_c);
+
+    a_a = node_a->get_column("a");
+    b_a = node_b->get_column("a");
+    c_a = node_c->get_column("a");
+  }
+
+  std::shared_ptr<MockNode> node_a, node_b, node_c;
+  std::shared_ptr<AbstractCostEstimator> cost_estimator;
+  LQPColumnReference a_a, b_a, c_a;
+};
+
+TEST_F(DpCcpTest, Basic) {
```

Comment:
Only two tests for this rather complex algorithm? I'd like to see additional tests with multiple tables and additional predicates. These two seem too simple to me :)

Reply:
First, the bulk of the testing is done once EnumerateCCP is proven to work. DPccp is just a "thin" layer around it, conceptually.

Secondly, and more importantly, testing DPccp means making assumptions about a cost model and a cardinality estimator and will thus become quite messy and will need to be changed every time these two are changed. I don't want anyone having to fix 10 DPccp tests everytime they change the cost model or the cardinality estimation.

Anyway, long story short, I'll think about adding more tests, but I might conclude that the current coverage is a good compromise.